### ISSUE TITLE
davmail has frequent wake-ups when idle

### ISSUE THREAD

**Issue by jdchristensen** [mguessan/davmail#199]
I'm using Ubuntu's davmail 5.5.1.3299-5 package on a laptop.  I have davmail configured for SMTP, but nothing else (no IMAP).  So davmail is doing nothing, except when I send an outgoing message.  However, powertop reports that davmail has 20 to 30 wake-ups per second.  Is davmail using a busy loop?  Can it be replaced by something that waits until it receives a connection?

I'm running davmail in server mode, so there is no icon.

My .davmail.properties file is [here](https://gist.github.com/jdchristensen/f01ed10eef54b2455beacbade196c301)

**Comment by mguessan** [comment]
This may be related to another issue I just fixed: with HttpClient 4 DavMail was way too aggressive on in idle/expired connections polling.
=> I changed code to poll less frequently and replaced the idle connection evictor thread with a scheduler

**Comment by jdchristensen** [comment]
The issue I'm seeing seems to involve a 50ms wait in a busy loop when davmail is only used for outgoing SMTP connections.  If you think it might be related, I'll test the latest version.  Will it be in the next Appveyor trunk build?  How often do they run?

**Comment by jdchristensen** [comment]
More precisely, there are two threads with wake-ups.  One has close to 20 per second, and the other has around 3.3 per second.  So they are at intervals of 50ms and ~300ms.

**Comment by mguessan** [comment]
Yes it might be related, and the appveyor build should always be up to date

**Comment by jdchristensen** [comment]
I just tried the latest Appveyor Debian package (on my Ubuntu system), and I still see two threads with 20 and 3.3 wake-ups per second.

**Comment by jdchristensen** [comment]
I saw commit 6a53540 from March, so I decided to test this again.  I just installed the latest Appveyor Debian package on my Ubuntu 22.04 system, and the two processes are still showing the same number of wake-ups.  On an idle system, with chrome running with dozens of tabs open, powertop estimates that davmail is the process that consumes the most battery.  And I'm only using it for outgoing SMTP, so it shouldn't need to be checking for anything until my client connects to it.  Any idea where this busy wait is?

**Comment by jdchristensen** [comment]
Just wondering if there has been any further thought into what is causing the wakeups.  I'm still getting them with 6.7.0p1-4072-1 on Ubuntu 24.04, and powertop indicates that this is using roughly 0.5W of power.

...Time passes...  I've now got systemd set-up to start davmail on demand, and shut it down after it is idle for 60s, which is perfect for my smtp-only use.  But it would be better if davmail waited without a busy loop.

**Comment by mguessan** [comment]
First thing to understand is if this is an issue with *all* java applications or just DavMail.

If this is a JDK level issue we can't do much about it.

Need more information on *which* thread triggers the wake-ups.

A tool like https://www.noureddine.org/research/joular/joularjx may help.

**Comment by jdchristensen** [comment]
It would also be helpful to know if others can reproduce this.  At least on linux, it's easy to run powertop and find out immediately if you get the wakeups.  @mguessan , if you are able to reproduce it, that would be helpful, as I'm not sure I'm going to be able to figure out joularjx.  OTOH, powertop does seem to give information at the level of threads.

BTW, my estimate of 0.5W is based not just on the davmail line shown by powertop, but also on the increase to the tick_nohz_highres_handler wake-ups, which use a fair bit of power.

In case it helps, here's the output of davmail when it starts:
```
Using system provided JRE
openjdk version "21.0.11" 2026-04-21
OpenJDK Runtime Environment (build 21.0.11+10-1-24.04.2-Ubuntu)
OpenJDK 64-Bit Server VM (build 21.0.11+10-1-24.04.2-Ubuntu, mixed mode, sharing)
Enable OpenJFX
```

**Comment by mguessan** [comment]
I don't have a native linux box, only virtual machines, no idea so far how to reproduce your issue, still not sure if this is a native OpenJDK issue or something specific to DavMail.

Maybe start with the obvious high level settings:

- do not start the listeners you don't use
```
davmail.popPort=0
davmail.imapPort=0
davmail.ldapPort=0
davmail.caldavPort=0
```
- ensure DavMail running in server mode:
```
davmail.server=true
```

If you can determine at least a thread name (as visible from java) we can start from there, without this information nothing can be done.

**Comment by jdchristensen** [comment]
> I don't have a native linux box, only virtual machines, no idea so far how to reproduce your issue, still not sure if this is a native OpenJDK issue or something specific to DavMail.
> 
> Maybe start with the obvious high level settings:
> 
> * do not start the listeners you don't use

Confirmed.

> * ensure DavMail running in server mode:

I start it with `davmail -server`, so I think this is covered.

> If you can determine at least a thread name (as visible from java) we can start from there, without this information nothing can be done.

I asked Claude for help with this, based on the thread ID for the busiest thread (the one that wakes up 20 times per second, even when davmail is idle), and it said:

> The powertop wakeup source reporting ~20 wakeups/second is the **"VM Periodic Task"** thread (a.k.a. HotSpot's WatcherThread) of the DavMail JVM process. This thread is a JVM internal — it runs housekeeping tasks (GC monitoring, memory service polling, etc.) on a fixed interval and cannot be controlled by application code.
> 
> The flag `-XX:GuaranteedSafepointInterval=0` (which is sometimes suggested for this) had no effect: wakeup rate remained ~26/second after applying it.
> 
> This is inherent to any long-running HotSpot/OpenJDK application and is not specific to DavMail. The only way to eliminate it would be GraalVM native-image compilation.

Then I asked it about the thread that wakes up 4 times per second when davmail is idle, and it said:

> This one is also a DavMail JVM thread — **"Monitor Deflation"** (TID 557482, Tgid 557433, which is a separate DavMail instance from the first one). 
> 
> The Monitor Deflation thread handles deflating Java object monitors (synchronized blocks) that are no longer in heavy use. It's another JVM internal thread, same situation as before — HotSpot creates it, application code can't control it, and you'd see it in any OpenJDK application.
> 
> 4 wakeups/second is quite low for this thread; it wakes up when the JVM decides monitor deflation is needed, so its rate varies with application activity. Nothing actionable here either.

I don't know enough about JVMs to know if Claude is correct here, so I asked it to write a java program that sleeps and does nothing else, and it had exactly 20 wake-ups per second.  So it does seem inherent to the JVM, which is surprising to me.  So I guess this can be closed!  Thanks for pointing out that I should investigate further.

**Comment by mguessan** [comment]
Thank you for your feedback, confirms this is internal OpenJDK behaviour.
